libs/s25main/mapGenerator/HeadQuarters.h

```diff
@@ -42,45 +43,71 @@ namespace rttr { namespace mapGenerator {
      *
      * @param map map to search for suitable HQ positions
      * @param area area within the HQ position should be
+     * @param playerDistanceToMountains desired player distance to mountain
      *
      * @return all suitable HQ positions witihn the specified area (or the entire map if the area is empty).
      * @throw runtime_error
      */
     template<class T_Container>
-    std::vector<MapPoint> FindHqPositions(const Map& map, const T_Container& area)
+    std::vector<MapPoint> FindHqPositions(const Map& map, const T_Container& area, unsigned playerDistanceToMountains)
     {
         if(area.empty())
         {
             throw std::runtime_error("Could not find any valid HQ position!");
         }
 
-        const auto isValid = [](const MapPoint& pt) { return pt.isValid(); };
-        const auto isObstacle = [&map](MapPoint point) {
-            return map.textureMap.Any(point, [](auto t) { return !t.Is(ETerrain::Buildable); });
+        const auto isObstacle = [&map](const MapPoint& pt) {
+            return map.textureMap.Any(pt, [](auto t) { return !t.Is(ETerrain::Buildable); });
         };
-        std::vector<MapPoint> hqs;
-        std::copy_if(map.hqPositions.begin(), map.hqPositions.end(), std::back_inserter(hqs), isValid);
 
-        // Quality of a map point as one player's HQ is a mix of:
-        // 1. distance to other players' HQs (higher = better)
-        // 2. distance to any other obstacle e.g. mountain & water (higher = better)
-        NodeMapBase<unsigned> potentialHqQuality = DistancesTo(hqs, map.size);
+        // Ensure possible HQ positions are far enough from any obstacles
         const auto obstacleDistance = DistancesTo(map.size, isObstacle);
-        const auto minDistance = helpers::clamp(GetMaximum(obstacleDistance, area), 2u, 4u);
+        const auto minObstacleDistance = helpers::clamp(GetMaximum(obstacleDistance, area), 2u, 4u);
+        const auto farFromObstacles = [&obstacleDistance, minObstacleDistance](const MapPoint& pt) {
+            return obstacleDistance[pt] >= minObstacleDistance;
+        };
+        std::vector<MapPoint> possiblePositions;
+        std::copy_if(area.begin(), area.end(), std::back_inserter(possiblePositions), farFromObstacles);
+
+        if(possiblePositions.empty())
+        {
+            return possiblePositions;
+        }
+
+        // Ensure HQ positions keep desired distance to mountains
+        const auto mountain = [&map](const MapPoint& pt) { return map.textureMap.Any(pt, IsMinableMountain); };
+        const auto mountainDistance = DistancesTo(map.size, mountain);
+        const auto desiredDistance =
+          std::max(GetMinimum(mountainDistance, possiblePositions), playerDistanceToMountains);
```

Comment:
Not sure about this, what is the intention? Without this and a higher min mountainDistance it will select the points which are as close as allowed by playerDistanceToMountains. With this it may choose ones which are even further. Is that what is intended?

BTW: Please don't overuse `auto`. My rule of thumb: If the type is on the RHS or complicated, use `auto` but in general prefer explicit types. Especially with mountainDistance and desiredDistance next to each other I'd suggest to rename the former `mountainDistances` (then you don't need the type to know it's a collection) and showing the type of desiredDistance as `unsigned`(?). Reason: You even assign this to `auto allowedMountainDistance` and then subtract 5 from that. To see if this is valid or underflows an unsigned you need to go to allowedMountainDistance, then desiredDistance, then playerDistanceToMountains and find out what GetMinimum returns.
I really like `auto` but there are situations (like this) where it makes reasoning and review hard

Reply:
Updated the code as mentioned above. I had a quick look through `autos`, will keep an eye on it in the future as well.